File: packages/sdk/go/http.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Generic net/http middleware: extracts the wire envelope and binds CaracalContext per request.

package sdk

import "net/http"

// Middleware returns an http.Handler middleware that binds a CaracalContext
// for each inbound request from envelope headers.
func (c *Caracal) Middleware(next http.Handler, opts ...RootOptions) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx, err := c.BindFromRequest(r.Context(), r, opts...)
		if err != nil {
			http.Error(w, "missing bearer token", http.StatusUnauthorized)
			return
		}
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

File: packages/sdk/ts/src/http.ts
/*
 * Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * HTTP server middleware that binds Caracal context after a verifier boundary.
 */

import { Caracal, type RootOptions } from './client.js'

export interface IncomingLike {
  headers: Record<string, string | string[] | undefined>
}

export type FastifyRequestLike = IncomingLike

export type ConnectMiddleware = (req: IncomingLike, res: unknown, next: (err?: unknown) => void) => void

export function caracalContextMiddleware(caracal: Caracal, opts: RootOptions = {}): ConnectMiddleware {
  return (req, _res, next) => {
    caracal
      .bindFromHeaders(
        req.headers,
        async () => {
          next()
        },
        opts,
      )
      .catch(next)
  }
}

export function caracalFastifyHook(caracal: Caracal, opts: RootOptions = {}) {
  return async (req: IncomingLike) => {
    await caracal.bindFromHeaders(req.headers, async () => undefined, opts)
  }
}
